File: ios/locus/Sources/locus/Core/TrackingRecoveryPolicy.swift
import Foundation

enum TrackingRecoveryAction: Equatable {
    case none
    case keepRunning
    case waitForConfig
    case clearDesiredState
    case stopRuntime
    case start
}

func decideTrackingRecovery(
    trackingDesired: Bool,
    runtimeEnabled: Bool,
    configStatus: PersistedConfigStatus,
    hasAlwaysAuthorization: Bool,
    locationServicesEnabled: Bool,
    stopOnTerminate: Bool
) -> TrackingRecoveryAction {
    if !trackingDesired && runtimeEnabled { return .stopRuntime }
    if !trackingDesired { return .none }
    if !hasAlwaysAuthorization || !locationServicesEnabled { return .clearDesiredState }
    if runtimeEnabled { return .keepRunning }
    if configStatus == .corrupt { return .waitForConfig }
    if stopOnTerminate { return .clearDesiredState }
    if configStatus == .absent { return .clearDesiredState }
    return .start
}

File: android/src/main/kotlin/dev/locus/core/TrackingRecoveryPolicy.kt
package dev.locus.core

internal enum class TrackingRecoveryOrigin {
    FLUTTER_ENGINE,
    FOREGROUND_SERVICE,
}

internal enum class TrackingRecoveryAction {
    NONE,
    KEEP_RUNNING,
    WAIT_FOR_CONFIG,
    CLEAR_DESIRED_STATE,
    STOP_RUNTIME,
    START,
}

internal fun decideTrackingRecovery(
    trackingDesired: Boolean,
    runtimeEnabled: Boolean,
    configStatus: PersistedConfigStatus,
    hasPermission: Boolean,
): TrackingRecoveryAction = when {
    !trackingDesired && runtimeEnabled -> TrackingRecoveryAction.STOP_RUNTIME
    !trackingDesired -> TrackingRecoveryAction.NONE
    !hasPermission -> TrackingRecoveryAction.CLEAR_DESIRED_STATE
    runtimeEnabled -> TrackingRecoveryAction.KEEP_RUNNING
    configStatus == PersistedConfigStatus.CORRUPT -> TrackingRecoveryAction.WAIT_FOR_CONFIG
    configStatus == PersistedConfigStatus.ABSENT -> TrackingRecoveryAction.CLEAR_DESIRED_STATE
    else -> TrackingRecoveryAction.START
}
